pydantic_ai_slim/pydantic_ai/_function_schema.py

```
@@ -154,7 +154,7 @@ def function_schema(  # noqa: C901
             if takes_ctx and index == 0:
                 # should be the `context` argument, skip
                 continue
-            # TODO warn?
+            warnings.warn(f'Parameter {p.name!r} is missing a type annotation, defaulting to Any', UserWarning)
```

Ruff (0.15.18)

[warning] No explicit `stacklevel` keyword argument found

Set `stacklevel=2`

(B028)